Object Detector Abstraction

NateMeyer opened the pull request:
This is an initial change to abstract the edgtpu module to a more generic Object Detector.  The idea is to be able to add different inference runtimes into the Detectors folder, add a new DetectorTypeEnum, and an else-if in the LocalObjectDetector init function to setup the new detector type.

The object_detection module keeps the multi-process logic to run the detector asynchronously and apply the detection thresholds.  All the detector modules are required to do is setup the runtime, load the model, and process a tensor (detect_raw).

Future work to build on this:
- Add an OpenVino detector module
- Add a RKNPU/RKNN detector module
- Possibly add a config for a "plugin" folder that will be scanned at runtime for user-supplied detector implementations?

NateMeyer committed:
Refactor EdgeTPU and CPU model handling to detector submodules.

NateMeyer committed:
Fix selecting the correct detection device type from the config

NateMeyer committed:
Remove detector type check when creating ObjectDetectProcess

NateMeyer committed:
Fixes after rebasing to 0.11

NateMeyer committed:
Add init file to detector folder

NickM-27 commented on frigate/object_detection.py:
```diff
@@ -24,45 +26,27 @@ def detect(self, tensor_input, threshold=0.4):
-                )
-                raise
+        if det_type == DetectorTypeEnum.edgetpu:
+            self.detectApi = EdgeTpuTfl(det_device=det_device, model_path=model_path)
```
```suggestion
            self.detect_api = EdgeTpuTfl(det_device=det_device, model_path=model_path)
```

NickM-27 commented on frigate/object_detection.py:
```diff
@@ -24,45 +26,27 @@ def detect(self, tensor_input, threshold=0.4):
-
-        self.tensor_input_details = self.interpreter.get_input_details()
-        self.tensor_output_details = self.interpreter.get_output_details()
+            self.detectApi = CpuTfl(model_path=model_path, num_threads=num_threads)
```
```suggestion
            self.detect_api = CpuTfl(model_path=model_path, num_threads=num_threads)
```

NickM-27 commented on frigate/object_detection.py:
```diff
@@ -79,31 +63,7 @@ def detect(self, tensor_input, threshold=0.4):
-            ]
-
-        return detections
+        return self.detectApi.detect_raw(tensor_input=tensor_input)
```
```suggestion
        return self.detect_api.detect_raw(tensor_input=tensor_input)
```

NickM-27 commented:
Thanks, this will be a good basis for #3626

NateMeyer committed:
Rename to detect_api

Co-authored-by: Nicolas Mowen <nickmowen213@gmail.com>

NickM-27 commented:
It would be nice to add some tests for this new module. Might be difficult / not a good idea to test actual detection, but at least testing that the correct detector is used when sending config metadata would be good.

NateMeyer commented:
> It would be nice to add some tests for this new module. Might be difficult / not a good idea to test actual detection, but at least testing that the correct detector is used when sending config metadata would be good.

Shouldn't be too hard if we can mock the edgetpu detector inside a unit test?  I'll take a look at that.

NickM-27 commented:
Yeah that's what I was thinking

NateMeyer committed:
Add unit test for LocalObjectDetector class

NateMeyer commented:
As I am working on an OpenVino detector using this branch, I'm finding it would be helpful if the configuration for a model were more descriptive.  Has there been any work to adjust the shape of the input/output tensors for different models?

NateMeyer commented:
I see some work in #1022, #3099, and #2548. Was there any preference on how any of these handled different models?

blakeblackshear commented:
I would prefer that each model implementation defines it's own function to prepare the input data in shared memory and then translates the outputs into a common format. Keep in mind that the detection process is extremely sensitive to small increases in processing time. When using a Coral, an extra copy of a frame in memory is enough to increase the total inference times by a significant percentage.

NateMeyer committed:
Add configuration for model inputs
Support transforming detection regions to RGB or BGR.
Support specifying the input tensor shape.  The tensor shape has a standard format ["BHWC"] when handed to the detector, but can be transformed in the detector to match the model shape using the model  input_tensor config.

NateMeyer committed:
Add documentation for new model config parameters

NateMeyer committed:
Add input tensor transpose to LocalObjectDetector

NateMeyer commented:
Looks like the input tensor shape can be transformed without a memory copy.  This can be handled by the detector, so I changed the "model_shape" to be the whole model config in most places.  The one place I needed to use the model config in video.py is to set the correct colorspace when we convert from YUV.  We'll only want to do this conversion once, so we need to know the correct colorspace for the model input at that time.

NickM-27 reviewed (approved):
Generally looks good to me, will defer to Blake on the model specifics

blakeblackshear commented:
I haven't had time to review this in detail yet, but Frigate+ models will be trained on the YUV colorspace, so don't make any assumptions about colorspace here.

NateMeyer commented:
The default colorspace should remain RGB as was used for the current default model and there is an option for `YUV` that will only crop/resize and skip the colorspace transformation call.  I was finding a mix of RGB and BGR in the Intel Open Model Zoo.

NickM-27 commented:
Please update this PR to point to `dev` branch

NickM-27 commented on frigate/object_detection.py:
```diff
@@ -23,46 +25,47 @@ def detect(self, tensor_input, threshold=0.4):
-                    "No EdgeTPU was detected. If you do not have a Coral device yet, you must configure CPU detectors."
-                )
-                raise
+        if model_config:
```
Is this supposed to be `self.model_config.input_tensor` ?

self.model_config will always be non-null since it is built by default even if a user doesn't define it in their config

NateMeyer replied:
I see there's a couple places where LocalObjectDetector is instantiated without passing in a model.  This check will save us from an error on dereferencing `model_config` if it is `None` and the `tensor_transform()` function will do another check on the argument.

NickM-27 replied:
Yeah I just missed those other locations as I was just looking in app

NateMeyer committed:
Change the model input tensor config to use an enumeration

NateMeyer commented:
I rebased and pushed one more commit for the model input config.  I switched it to an enum to specify NCHW vs NHWC tensor shapes instead of an array of arbitrary format.  I thought that other shapes might be out there, but I haven't come across any besides these two yet.  Best to keep it easier to configure for now.

NateMeyer committed:
Updates for model config documentation

triptec commented:
@NateMeyer this looks great! May I ask if you have any example config/model where you use this new functionality?

NateMeyer commented:
> @NateMeyer this looks great! May I ask if you have any example config/model where you use this new functionality?

Hey. The config changes should be included in the documentation updates. I implemented an OpenVino detector using this abstraction in #3768
